src/Perl6/Metamodel/CurriedRoleHOW.nqp

```diff
@@ -79,77 +79,94 @@ class Perl6::Metamodel::CurriedRoleHOW
         nqp::settypecheckmode($type, 2);
     }
 
-    method parameterize_roles($obj) {
-        my @pos_args;
-        nqp::push(@pos_args, $obj);
-        for @!pos_args {
-            nqp::push(@pos_args, $_);
+    # A curried role may be bound to form a role other than its origin, e.g.  a
+    # parametric role group and the candidate we're not quite ready to fetch at
+    # construction time. If we're still not ready yet whenever we need it, the
+    # binding is ourself for now. If the origin lacks a means of producing a
+    # binding, we assume it's its own by default.
+    method bind($obj) {
+        if $!binding =:= NQPMu {
+            if nqp::can($!curried_role.HOW, 'bind') {
+                my $binding := $!curried_role.HOW.bind($!curried_role, $obj);
+                $!binding := $binding unless nqp::decont($binding) =:= $obj;
+                $binding
+            }
+            else {
+                $!binding := $!curried_role
+            }
         }
-        if nqp::istype($!curried_role.HOW, Perl6::Metamodel::ParametricRoleGroupHOW) {
-            $!candidate := $!curried_role.HOW.select_candidate($!curried_role, @pos_args, %!named_args);
-            my $candidate-how := $!candidate.HOW;
+        else {
+            $!binding
+        }
+    }
 
-            self.set_language_revision($obj, $candidate-how.language-revision($!candidate));
+    method parameterize_roles($obj) {
+        if self.archetypes.generic {
+            $!is_complete := 1;
```

Comment:
It is kind of a good idea for optimization, but it breaks typechecking of partially parameterized roles:

```raku
role R1[::T] { }
role R2[::T1, ::T2] does R1[::T1] { } 
role R3[::T] does R2[Int, T] { }
say R3.^candidates.head ~~ R1[Int];
```

This is _True_ on the current master but _False_ on your branch. Perhaps this makes sense when all arguments are generic, but what if there are too many of them?